### ISSUE TITLE
etcd_test fails with an assertion in gRPC

### ISSUE THREAD

**Issue by prasad-joshi** [etcd-cpp-apiv3/etcd-cpp-apiv3#5]
I had same problems during compilation as mentioned on https://github.com/nokia/etcd-cpp-apiv3/issues/2

Fixed them by changing `cmake` file like this 

```
prasad@prasad-Inspiron-5567:~/sources/nokia/etcd-cpp-apiv3$ git diff
diff --git a/tst/CMakeLists.txt b/tst/CMakeLists.txt
index e0ce264..53a515b 100644
--- a/tst/CMakeLists.txt
+++ b/tst/CMakeLists.txt
@@ -1,9 +1,12 @@
-find_path(CATCH_INCLUDE_DIR NAMES catch.hpp)
+find_path(CATCH_INCLUDE_DIR NAMES catch.hpp PATHS ${CMAKE_SOURCE_DIR})
 include_directories(${CATCH_INCLUDE_DIR})
 
 add_executable(etcd_test EtcdTest.cpp EtcdSyncTest.cpp WatcherTest.cpp)
 set_property(TARGET etcd_test PROPERTY CXX_STANDARD 11)
 
-target_link_libraries(etcd_test etcd-cpp-api)
+target_link_libraries(etcd_test
+       etcd-cpp-api
+       pthread
+)
 
 add_test(etcd_test etcd_test)

```
After making above changes the library compiles however, `make test` fails

```
root@ubuntu16:~/workspace/nokia/etcd-cpp-apiv3/build# make test
Running tests...
Test project /home/prasad/workspace/nokia/etcd-cpp-apiv3/build
    Start 1: etcd_test
1/1 Test #1: etcd_test ........................***Failed    0.76 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.78 sec

The following tests FAILED:
	  1 - etcd_test (Failed)
Errors while running CTest
Makefile:105: recipe for target 'test' failed
make: *** [test] Error 8

```

It seems like an assertion in gRPC failed.

```
root@ubuntu16:~/workspace/nokia/etcd-cpp-apiv3/build# ./tst/etcd_test  
E0522 13:22:31.865897860   10686 channel_cc.cc:136]          assertion failed: GRPC_CALL_OK == grpc_call_start_batch(call->call(), cops, nops, ops, nullptr)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
etcd_test is a Catch v1.5.4 host application.
Run with -? for options

-------------------------------------------------------------------------------
wait for a value change
-------------------------------------------------------------------------------
/home/prasad/workspace/nokia/etcd-cpp-apiv3/tst/EtcdTest.cpp:277
...............................................................................

/home/prasad/workspace/nokia/etcd-cpp-apiv3/tst/EtcdTest.cpp:277: FAILED:
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

===============================================================================
test cases:  14 |  13 passed | 1 failed
assertions: 108 | 107 passed | 1 failed

```

**Comment by zhangwuliang** [comment]
I have also encountered this problem, Have you solved it? How to solve?

**Comment by bjeremy** [comment]
I also have the same problem... Has anyone found out what the reason is?

**Comment by siredmar** [comment]
Same issue here. Testing with the modifications given by bjeremy didn't gave better results. Is this project still maintained?

**Comment by fleapapa** [comment]
i made the following changes (thanks to bjeremy):
```
--- a/v3/src/AsyncWatchAction.cpp
+++ b/v3/src/AsyncWatchAction.cpp
@@ -11,7 +11,9 @@ etcdv3::AsyncWatchAction::AsyncWatchAction(etcdv3::ActionParameters param)
 {
   isCancelled = false;
   stream = parameters.watch_stub->AsyncWatch(&context,&cq_,(void*)"create");
-
+void* got_tag;
+bool ok = false;
+cq_.Next(&got_tag, &ok);
   WatchRequest watch_req;
   WatchCreateRequest watch_create_req;
   watch_create_req.set_key(parameters.key);
@@ -79,6 +81,7 @@ void etcdv3::AsyncWatchAction::waitForResponse(std::function<void(etcd::Response
     if(got_tag == (void*)"writes done")
     {
       isCancelled = true;
+      break;
     }
     else if(got_tag == (void*)this) // read tag
     {
```

Then, test happened to **all pass** (really? kinda not sure b/c of the `E0119 `  :)
```
$ ./etcd_test 
print response called
create 42
Previous value: 
print response called
set 43
Previous value: 42
print response called
delete 
Previous value: 43
print response called
create 44
Previous value: 
E0119 00:36:00.334081015    9765 backup_poller.cc:110]       run_poller: {"created":"@1547886960.334040032","description":"Shutting down timer system","file":"src/core/lib/iomgr/timer_generic.cc","file_line":704}
print response called
create 42
Previous value: 
print response called
set 43
Previous value: 42
===============================================================================
All tests passed (217 assertions in 21 test cases)
```

In case you have any formal or better fix, please don't hesitate to share it. Thanks!

**Comment by chenyicong** [comment]
@siredmar：
actually, you can change the code which provided by bjeremy to below：
etcdv3::AsyncWatchAction::AsyncWatchAction(...)
{
isCancelled = false;
stream = parameters.watch_stub->AsyncWatch(&context,&cq_,(void*)"create");
....
gpr_timespec deadline;
deadline.clock_type = GPR_TIMESPAN;
deadline.tv_sec = 0;
deadline.tv_nsec = 10000000;
auto status = cq_.AsyncNext(&got_tag, &ok, deadline);

- if (status != CompletionQueue::SHUTDOWN) ----delete
+ if (status == CompletionQueue::GOT_EVENT)  ---add
{
stream->Write(watch_req, (void*)"write");
stream->Read(&reply, (void*)this);
}
}

because you can see GRPC code， only the GOT_EVENT staus can do any stream action： 
  /// Tri-state return for AsyncNext: SHUTDOWN, GOT_EVENT, TIMEOUT.
  enum NextStatus {
    SHUTDOWN,   ///< The completion queue has been shutdown and fully-drained
    GOT_EVENT,  ///< Got a new event; \a tag will be filled in with its
                ///< associated value; \a ok indicating its success.
    TIMEOUT     ///< deadline was reached.
  };

**Comment by sighingnow** [comment]
Has been fixed now, by #16.

**Comment by sighingnow** [comment]
Thanks everyone who are interested in this library!
